### ISSUE TITLE
Contour lines left off layer if there are too many

### ISSUE THREAD

**Issue by davehemming** [KNMI/adaguc-server#140]
When adding lots of custom ContourLine's there seems to be a limit to how many the server will display.  If I go over that limit some lines that should show up do not.  

For example, here I am trying to represent the difference in height between two levels of pressure, the height of the 500mb pressure level minus height of 1000mb level (thickness).  I am using the [GFS forecast](https://nomads.ncep.noaa.gov/dods/gfs_0p25), specifically the 'hgtprs' variable.  Here is my style:

```
<Style name="thickness">
    <Legend>not_used</Legend>
    <ContourLine width=".6" linecolor="#FF0000" classes="5760" />
    <ContourLine width=".6" linecolor="#FF7700" classes="5520,5540,5560,5580,5600,5620,5640,5660,5680,5700,5720" />
    <ContourLine width=".6" linecolor="#00FFFF" textcolor="#007979" textformatting="2100m" classes="5480" />
    <ContourLine width=".6" linecolor="#00FFFF" textcolor="#007979" textformatting="1800m" classes="5440" />
    <ContourLine width=".6" linecolor="#0000FF" textcolor="#0000FF" textformatting="1500m" classes="5400" />
    <ContourLine width=".6" linecolor="#0000FF" textcolor="#0000FF" textformatting="1200m" classes="5360" />
    <ContourLine width=".6" linecolor="#DD00FF" textcolor="#DD00FF" textformatting="900m" classes="5320" />
    <ContourLine width=".6" linecolor="#DD00FF" textcolor="#DD00FF" textformatting="600m" classes="5280" />
    <ContourLine width=".6" linecolor="#BB00FF" textcolor="#BB00FF" textformatting="300m" classes="5240" />
    <ContourLine width=".6" linecolor="#BB00FF" textcolor="#BB00FF" textformatting="Sea Level" classes="5200" />
    <ContourLine width=".6" linecolor="#BB00FF" classes="5100,5120,5140,5160,5180" />

    <RenderMethod>contour</RenderMethod>
    <NameMapping name="contour" title="Thickness contour" abstract="Contourlines on a transparent background" />
</Style>
```

When I display this the 'Sea Level' contour line is not shown.  If I remove a few of these ContourLine's it shows up again suggesting there is a hard limit on how many ContourLine's the server will show.  Is it a setting or something on the server?

**Comment by maartenplieger** [comment]
Hi Dave,

Thanks for your questions (Also the other one). I will have to check these. Yes there might be a limit, it was unforseen that someone would hit the limit. If so we can increase that.

But did you know that you can also give it an interval, so just contour at any 100th value? (https://dev.knmi.nl/projects/adagucserver/wiki/ContourLine)

Your style config below seems incomplete, I cannot see it. Can you re-share?

If you want, I am quite happy to have a chat/call with you as well to understand what you are doing with adaguc-server. It is nice to hear that you are still using it.

Cheers,
Maarten
________________________________________
Van: davehemming <notifications@github.com>
Verzonden: donderdag 27 augustus 2020 04:20:59
Aan: KNMI/adaguc-server
CC: Subscribed
Onderwerp: [KNMI/adaguc-server] Contour lines left off layer if there are too many (#140)

When adding lots of custom ContourLine's there seems to be a limit to how many the server will display. If I go over that limit some lines that should show up do not.

For example, here I am trying to represent the difference in height between two levels of pressure, the height of the 500mb pressure level minus height of 1000mb level (thickness). I am using the GFS forecast<https://nomads.ncep.noaa.gov/dods/gfs_0p25>, specifically the 'hgtprs' variable. Here is my style:

`<Style name="thickness">
not_used











    <RenderMethod>contour</RenderMethod>
    <NameMapping name="contour" title="Thickness contour" abstract="Contourlines on a transparent background" />
</Style>`


When I display this the 'Sea Level' contour line is not shown. If I remove a few of these ContourLine's it shows up again suggesting there is a hard limit on how many ContourLine's the server will show. Is it a setting or something on the server?

**Comment by davehemming** [comment]
Thanks for the speedy reply!

There is not much to this dataset, but here are the main bits:

```
<Legend name="not_used" type="colorRange">
    <palette index="0" red="0" green="0" blue="0" />
    <palette index="240" red="255" green="255" blue="255" />
</Legend>

<Style name="thickness">
    <Legend>not_used</Legend>
    <ContourLine width=".6" linecolor="#FF0000" classes="5760" />
    <ContourLine width=".6" linecolor="#FF7700" classes="5520,5540,5560,5580,5600,5620,5640,5660,5680,5700,5720" />
    <ContourLine width=".6" linecolor="#00FFFF" textcolor="#007979" textformatting="2100m" classes="5480" />
    <ContourLine width=".6" linecolor="#00FFFF" textcolor="#007979" textformatting="1800m" classes="5440" />
    <ContourLine width=".6" linecolor="#0000FF" textcolor="#0000FF" textformatting="1500m" classes="5400" />
    <ContourLine width=".6" linecolor="#0000FF" textcolor="#0000FF" textformatting="1200m" classes="5360" />
    <ContourLine width=".6" linecolor="#DD00FF" textcolor="#DD00FF" textformatting="900m" classes="5320" />
    <ContourLine width=".6" linecolor="#DD00FF" textcolor="#DD00FF" textformatting="600m" classes="5280" />
    <ContourLine width=".6" linecolor="#BB00FF" textcolor="#BB00FF" textformatting="300m" classes="5240" />
    <ContourLine width=".6" linecolor="#BB00FF" textcolor="#BB00FF" textformatting="Sea Level" classes="5200" />
    <ContourLine width=".6" linecolor="#BB00FF" classes="5100,5120,5140,5160,5180" />

    <RenderMethod>contour</RenderMethod>
    <NameMapping name="contour" title="Thickness contour" abstract="Contourlines on a transparent background" />
</Style>

<Layer type="database">
    <Name>thk</Name>
    <Title>thk</Title>
    <FilePath>/data/adaguc-data/noaa/GFS0P25/GFS0P25AU202008251800</FilePath>
    <Variable>thk</Variable>
    <Styles>thickness</Styles>
    <Dimension name="time" default="min">time</Dimension>
</Layer>
```

Yes I thought about using an interval instead of using classes on the ContourLine tag, but I need to restrict it to a range of values.  E.g. Between 5520 and 5720 display a contour line every 20th value with the colour #FF7700.  I was going to propose a range attribute which might be helpful for such use cases where you could have a start and end value, or a single value with a + or - so that it adds contour lines above or below it every nth step.

For the moment I have solved the problem by splitting it into two styles, and serving that to the front end web application which then combines them into a single layer.  Like so:

```
<Legend name="not_used" type="colorRange">
    <palette index="0" red="0" green="0" blue="0" />
    <palette index="240" red="255" green="255" blue="255" />
</Legend>

<Style name="thickness1">
    <Legend>not_used</Legend>
    <ContourLine width=".6" linecolor="#FF0000" classes="5760" />
    <ContourLine width=".6" linecolor="#FF7700" classes="5520,5540,5560,5580,5600,5620,5640,5660,5680,5700,5720" />
    <ContourLine width=".6" linecolor="#BB00FF" classes="5100,5120,5140,5160,5180" />
    <RenderMethod>contour</RenderMethod>
    <NameMapping name="contour" title="Thickness contour" abstract="Contourlines on a transparent background" />
</Style>

<Style name="thickness2">
    <Legend>not_used</Legend>
    <ContourLine width=".6" linecolor="#00FFFF" textcolor="#007979" textformatting="2100m" classes="5480" />
    <ContourLine width=".6" linecolor="#00FFFF" textcolor="#007979" textformatting="1800m" classes="5440" />
    <ContourLine width=".6" linecolor="#0000FF" textcolor="#0000FF" textformatting="1500m" classes="5400" />
    <ContourLine width=".6" linecolor="#0000FF" textcolor="#0000FF" textformatting="1200m" classes="5360" />
    <ContourLine width=".6" linecolor="#DD00FF" textcolor="#DD00FF" textformatting="900m" classes="5320" />
    <ContourLine width=".6" linecolor="#DD00FF" textcolor="#DD00FF" textformatting="600m" classes="5280" />
    <ContourLine width=".6" linecolor="#BB00FF" textcolor="#BB00FF" textformatting="300m" classes="5240" />
    <ContourLine width=".6" linecolor="#BB00FF" textcolor="#BB00FF" textformatting="Sea Level" classes="5200" />

    <RenderMethod>contour</RenderMethod>
    <NameMapping name="contour" title="Thickness contour" abstract="Contourlines on a transparent background" />
</Style>

<Layer type="database">
    <Name>thk</Name>
    <Title>thk</Title>
    <FilePath>/data/adaguc-data/noaa/GFS0P25/GFS0P25AU202008251800</FilePath>
    <Variable>thk</Variable>
    <Styles>thickness1,thickness2</Styles>
    <Dimension name="time" default="min">time</Dimension>
</Layer>
```

Prior to the above solution I tried to use the DataPostProc tag with the include_layer attribute but was not able to get it to work. 
 The reason I need so many ContourLine's is I need different colours and different labels depending on the value they represent.

Yes would be interested to chat sometime, that would be great if you could spare a moment.  I think you might like to see how I am using ADAGUC, I'm sure you will have some constructive feedback to share.

**Comment by maartenplieger** [comment]
Hi Dave,

There is indeed a limit on the amount of contour lines. It is currently set to 16. This to identify the different lines in memory in a grid which size equals the GetMap request. I believe we use a byte from which the first 4 are to identify the contour line number. Hence the 16 limit. We can use a unsigned short and have a limit of 256 contour lines.

It's done in https://github.com/KNMI/adaguc-server/blob/master/adagucserverEC/CImgWarpBilinear.cpp

So we can increase from 16 to 256, but there will still be a limit of 256. Would this be OK?

**Comment by davehemming** [comment]
Oh nice, thanks for the explanation and link to the file in question, that's very interesting.  Yes 256 is plenty, that would be fantastic 👍

**Comment by davehemming** [comment]
Hi I tried changing the lines:

```
#define CONTOURDEFINITIONLOOKUPTYPE unsigned char
#define CONTOURDEFINITIONLOOKUPLENGTH 16
```

to:

```
#define CONTOURDEFINITIONLOOKUPTYPE unsigned short
#define CONTOURDEFINITIONLOOKUPLENGTH 256
```

Then rebuilding the docker container but didn't seem to make a difference.  Also I did a project wide search for the method 'checkIfContourRequired' in which 'CONTOURDEFINITIONLOOKUPLENGTH' is referenced but I could not see any usage of this method anywhere in the codebase?  Surely I'm missing something.

Sorry when it comes to C++ I'm out of my depth here.  I've got some ugly code on the frontend that is combining two separate layers to hack around this and I'd like to remove it if I can.

**Comment by maartenplieger** [comment]
Thanks Dave, I will have a look now

**Comment by maartenplieger** [comment]
Hi Dave, can you check this out: https://github.com/KNMI/adaguc-server/tree/140-allow-more-contourlines

**Comment by davehemming** [comment]
Looks great thank you!

By the way in `compile.sh` I had to add the method call `clean` to line 40 to get the docker container to build otherwise I was  getting a directory not found error.

**Comment by maartenplieger** [comment]
Hi Dave, Fixed your issue in this branch. And the amount of possible contourline definitions is now 32.

**Comment by maartenplieger** [comment]
Resolved in version 2.5.1
